[CS] Use fixes to diagnose instance member on type (or vice versa) access

theblixguy opened the pull request:
This PR migrates instance member on type and type member on instance diagnostics handling to use the new diagnostics framework (fixes) and create more reliable and accurate diagnostics in such scenarios.

Commit:
[sema] [csdiag] adds FailureDiagnostic subclass

Commit:
[sema] [csfix] adds CSFix subclass

Commit:
[sema] [cs] record fix and emit diagnostic

Commit:
[sema] fix indent

Commit:
[sema] [csfix] fix indent

Commit:
[cs] fix indent

Commit:
[sema] fix indent and update some names

Commit:
[sema] add some comments

Commit:
[sema] use unviable result for diagnostics

Commit:
[sema] [cs] update diagnostics and handle both cases

Commit:
[sema] [cs] add comments and fix indent

Commit:
[test] add some tests

theblixguy commented:
cc @xedin @slavapestov

Commit:
[cs] [test] fix typo & update some names

slavapestov commented on test/decl/func/functions.swift:
```diff
@@ -179,3 +179,29 @@ func parentheticalInout2(_ fn: (((inout Int)), Int) -> ()) {
   fn(&value, 0)
 }
+
+class A {}
```
I think test/Constraints/members is a better place for these tests

theblixguy replied:
Thanks, I have moved it to that file now 👍

Commit:
[test] move tests to correct test file

Commit:
[sema] [cs] remove old code

Commit:
[sema] [cs] use tc.diagnose instead of emitDiagnostic

theblixguy commented:
> Great! I will take a closer look later today but one thing I see right away - please remove relevant code from CSDiag.cpp which is now obsolete.

Oh yeah, done :) I am not sure about this block of code: https://github.com/apple/swift/blob/master/lib/Sema/CSDiag.cpp#L1006L1022 should this also be handled in `diagnoseAsError()`?

Commit:
[sema] [cs] re-indent code

xedin commented on lib/Sema/CSDiag.cpp:
```diff
@@ -845,9 +845,6 @@ diagnoseTypeMemberOnInstanceLookup(Type baseObjTy,
   if (isa<EnumElementDecl>(member))
     Diag.emplace(diagnose(loc, diag::could_not_use_enum_element_on_instance,
                           memberName));
-  else
```
I expect new fix to diagnose all of the cases previously diagnosed by CSDiag for `UR_InstanceMemberOnType` as well as `UR_TypeMemberOnInstance`, so `diagnoseTypeMemberOnInstanceLookup` should be completely eliminated as well as portions of `diagnoseUnviableLookupResults`.

theblixguy replied:
Oh, I guess I’ll have to modify my class to accept more arguments, like `member`, just like those functions!

xedin replied:
Yeah, and some of the stuff you can get from constraint system too, since solution gets applied before diagnosing fixes.

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -1477,3 +1477,18 @@ bool MissingMemberFailure::diagnoseAsError() {
   return true;
 }
+
+bool AllowTypeOrInstanceMemberFailure::diagnoseAsError() {
```
Since "fix" always leads to a solution we should be able to always suggest a fix-it to correct situation along with diagnostic.

theblixguy replied:
That will be handled by existing diagnostic code once I moved it over. Do you think we need new diagnostics for some cases?

xedin replied:
Last time I checked not all of them produced a fix-it unfortunately, you just need to be aware of that...

xedin commented on lib/Sema/CSDiagnostics.h:
````diff
@@ -674,6 +674,37 @@ class MissingMemberFailure final : public FailureDiagnostic {
+///
+/// }
+/// ```
+
````
Nit: Unnecessary newline here.

xedin commented on lib/Sema/CSDiagnostics.h:
```diff
@@ -674,6 +674,37 @@ class MissingMemberFailure final : public FailureDiagnostic {
+        Name(memberName) {}
+
+  bool diagnoseAsError() override;
+
```
And here...

xedin commented on lib/Sema/CSFix.h:
```diff
@@ -493,6 +496,27 @@ class DefineMemberBasedOnUse final : public ConstraintFix {
+        BaseType(baseType), Name(member) {}
+
+  std::string getName() const override {
+    return "allow type or instance member";
```
It would be great if this message provided more information about what did happen exactly.

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -3344,6 +3344,7 @@ performMemberLookup(ConstraintKind constraintKind, DeclName memberName,
     bool hasInstanceMembers = false;
     bool hasInstanceMethods = false;
     bool hasStaticMembers = false;
+    bool isMetatype = baseTy->is<MetatypeType>();
```
Should be `isMetatypeBase` probably?

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -3699,10 +3701,9 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyMemberConstraint(
+	
   // If we found some unviable results, then fail, but without recovery.
-  if (!result.UnviableCandidates.empty())
+  if (!result.UnviableCandidates.empty() && !shouldAttemptFixes())
```
I think you can remove this check completely because we return `SolutionKind:Error` after `if (shouldAttemptFixes())` block already.

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -3825,8 +3826,41 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyMemberConstraint(
+    if (!result.UnviableCandidates.empty()) {
+      // Check if we have unviable candidates whose reason for rejection
+      // is either UR_InstanceMemberOnType or UR_TypeMemberOnInstance
+      auto unviable = [&](const std::pair<OverloadChoice,
```
`unviable` is not a very good name for this closure, just write it inline since it's a single use anyway.

theblixguy replied:
Yeah, I moved it out as the code looked a bit weird visually. I’ll revert it to being inline like before.

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -3825,8 +3826,41 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyMemberConstraint(
+        auto overloadChoices = SmallVector<OverloadChoice, 4>();
+        
+        for (auto &pair : result.UnviableCandidates) {
+          overloadChoices.push_back(pair.first);
```
Should `OverloadChoice` be moved to new vector?

theblixguy replied:
Yes as I can’t pass `UnviableCandidates` directly as it’s an array of `std::pair<...>`, so I need to create a vector to store the overload choices, unless there’s another way to extract them directly?

xedin replied:
Sorry, you can keep the array, what I mean is to use `std::move` to move elements over from unviable list, instead of copying :)

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -3825,8 +3826,41 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyMemberConstraint(
+          overloadChoices.push_back(pair.first);
+        }
+        
+        addOverloadSet(memberTy, ArrayRef<OverloadChoice>(overloadChoices), useDC,
```
I wounder why it's necessary to add `ArrayRef<OverloadChoice>` here? There is an `ArrayRef` constructor which takes `SmallVector` already as far as I remember...

theblixguy replied:
I was able to pass `overloadChoices` directly but it led to a crash: `Assertion failed: (decl), function operator(), file /Users/suyashsrijan/Documents/swift-src/swift/lib/Sema/ConstraintSystem.cpp, line 1404.`. 

Creating a `ArrayRef<OverloadChoice>` using it and then passing it fixed the issue - not sure why!

xedin replied:
Probably Overload decl didn’t get properly copy constructed or something, try moving them into new vector, let’s see if that fixes this issue

xedin reviewed (changes requested):
Overall I think this is very good start, but definitely needs more work...

Commit:
[csdiag] move old code to the new fix

Commit:
[csdiag] remove old code

Commit:
[csfix] update getName() message

Commit:
[cssimplify] Move overload choices instead of copying, etc

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -3825,8 +3822,41 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyMemberConstraint(
+        
+        // The fix was successful, so let's add it to the overload set
+        // and return the solution as solved
+        auto overloadChoices = SmallVector<OverloadChoice, 4>();
```
Nit: I haven't actually seen this style anywhere in our repository, might be worth to change it to `SmallVector<OverloadChoice, 4> choices;` instead.

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -1477,3 +1477,223 @@ bool MissingMemberFailure::diagnoseAsError() {
   return true;
 }
+
+bool AllowTypeOrInstanceMemberFailure::diagnoseAsError() {
```
Could you please run clang-format on this method again? It seems incorrectly indented...

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -1477,3 +1477,223 @@ bool MissingMemberFailure::diagnoseAsError() {
+      
+      if (TypeDC->getSelfNominalTypeDecl() == instanceTy->getAnyNominal()) {
+        if (propertyInitializer) {
+          tc.diagnose(loc, diag::instance_member_in_initializer,
```
Instead of using `TypeChecker::diagnose` directly please switch to `emitDiagnostic` which is already provided by `FailureDiagnostic`

Commit:
[cs] get member from getResolvedOverloadSets

xedin reviewed (commented):
I think this is looking much better, almost there!

Commit:
[csdiag] replace break with unreachable

Commit:
[csdiagnostics] replace tc.diagnose with emitDiagnostic

Commit:
[cs] rename overloadChoices to choices

Commit:
[cs] use getResolvedOverload() to get the member

Commit:
[cs] re-indent

theblixguy commented:
Thanks @xedin, can you take another look?

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -3425,8 +3426,9 @@ performMemberLookup(ConstraintKind constraintKind, DeclName memberName,
-      if ((isa<FuncDecl>(decl) && !hasInstanceMethods) ||
-          (!isa<FuncDecl>(decl) && !hasInstanceMembers)) {
+      if (((isa<FuncDecl>(decl) && !hasInstanceMethods)
+           || (!isa<FuncDecl>(decl) && !hasInstanceMembers))
```
Nit: this is not how clang-format indents operators - they are always left at the end of the line.

theblixguy replied:
Let me run it on all the files I have changed again!

xedin commented:
@theblixguy LGTM! Have you ran tests locally to check if anything broke?

xedin commented:
Although it's good to run validation test but it's not that interesting, try running regular tests via `-t` (also both together are `-T`).

Commit:
Merge branch 'master' into fix/SR-9415

Commit:
Merge branch 'master' into fix/SR-9415

Commit:
Merge branch 'master' into fix/SR-9415

theblixguy commented:
Thanks for the tip @xedin - I haven't been able to build successfully, it keeps failing:

```
make: *** [/Users/suyashsrijan/Documents/swift-src/build/Xcode-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/Darwin/Foundation/macosx/x86_64/Foundation.o] Error 1

** BUILD FAILED **


The following build commands failed:
	PhaseScriptExecution CMake\ Rules /Users/suyashsrijan/Documents/swift-src/build/Xcode-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/Darwin/Foundation/Swift.build/RelWithDebInfo/stdlib-public-Darwin-Foundation-macosx-x86_64-Foundation.o.build/Script-62403CE08F0A4F85BD83D6F6.sh
(1 failure)
./utils/build-script: fatal error: command terminated with a non-zero exit status 65, aborting
```

Scrolling up the console output, I can see:

```
Traceback (most recent call last):
  File "/Users/suyashsrijan/Documents/swift-src/swift/utils/line-directive", line 741, in <module>
    run()
  File "/Users/suyashsrijan/Documents/swift-src/swift/utils/line-directive", line 712, in run
    input = command.stdout.readline()
KeyboardInterrupt
make: *** [/Users/suyashsrijan/Documents/swift-src/build/Xcode-RelWithDebInfoAssert/swift-macosx-x86_64/RelWithDebInfo/lib/swift/macosx/x86_64/Foundation.swiftmodule] Error 1
```

I have tried deleting the entire src folder and re-checking out everything but that didn't fix it. I am invoking the build-script with the following parameters: `-x -c -t --release-debuginfo`. Do you have any ideas why the build could be failing? The console output doesn't provide the exact issue.

xedin commented:
@theblixguy Try this from your swift directory - `./utils/build-script -r -T --skip-build-benchmarks`

theblixguy commented:
@xedin Same issue:

> [1029/1216][ 84%][1006.494s] Compiling /Users/suyashsrijan/Documents/swift-src/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/core/macosx/x86_64/Swift.o
ninja: build stopped: subcommand failed.
./utils/build-script: fatal error: command terminated with a non-zero exit status 1, aborting

It fails on:

> FAILED: lib/swift/macosx/x86_64/Foundation.swiftmodule lib/swift/macosx/x86_64/Foundation.swiftdoc lib/swift/macosx/x86_64/Foundation.swiftinterface
cd /Users/suyashsrijan/Documents/swift-src/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/Darwin/Foundation && /usr/local/Cellar/cmake/3.12.4/bin/cmake -E remove -f /Users/suyashsrijan/Documents/swift-src/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./lib/swift/macosx/x86_64/Foundation.swiftmodule /Users/suyashsrijan/Documents/swift-src/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./lib/swift/macosx/x86_64/Foundation.swiftdoc /Users/suyashsrijan/Documents/swift-src/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./lib/swift/macosx/x86_64/Foundation.swiftinterface && /usr/bin/python /Users/suyashsrijan/Documents/swift-src/swift/utils/line-directive @/Users/suyashsrijan/Documents/swift-src/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/Darwin/Foundation/EDgmP.txt -- /Users/suyashsrijan/Documents/swift-src/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./bin/swiftc -emit-module -o /Users/suyashsrijan/Documents/swift-src/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./lib/swift/macosx/x86_64/Foundation.swiftmodule -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk -target x86_64-apple-macosx10.9 -resource-dir /Users/suyashsrijan/Documents/swift-src/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./lib/swift -F /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/../../../Developer/Library/Frameworks -O -g -D INTERNAL_CHECKS_ENABLED -D SWIFT_ENABLE_RUNTIME_FUNCTION_COUNTERS -I /Users/suyashsrijan/Documents/swift-src/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./lib/swift/macosx/x86_64 -module-cache-path /Users/suyashsrijan/Documents/swift-src/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./module-cache -no-link-objc-runtime -Xfrontend -enable-resilience -Xfrontend -enable-sil-ownership -Xfrontend -enable-mandatory-semantic-arc-opts -Xfrontend -enforce-exclusivity=unchecked -swift-version 5 -swift-version 4 -autolink-force-load -warn-swift3-objc-inference-complete -Xfrontend -verify-syntax-tree -Xllvm -sil-inline-generics -Xllvm -sil-partial-specialization -swift-version 5 -warn-implicit-overrides -Fsystem /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/PrivateFrameworks/ -module-link-name swiftFoundation -force-single-frontend-invocation -parse-as-library -emit-parseable-module-interface @/Users/suyashsrijan/Documents/swift-src/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/Darwin/Foundation/EDgmP.txt

theblixguy commented:
There are a bunch of warnings related to Hashable: `warning: 'Hashable.hashValue' is deprecated as a protocol requirement; conform type 'AffineTransform' to 'Hashable' by implementing 'hash(into:)' instead`

Commit:
Merge branch 'master' into fix/SR-9415

xedin commented:
Hm, I'm not sure what's up with that then... Anyway, let's just run it here.

xedin commented:
@swift-ci please smoke test

xedin commented:
@theblixguy Ah hm, looks like we'd need a guard for cases like this, it creates `DotSyntaxBaseIgnoredExpr` as a result because it just passes function type as a value...

xedin commented:
@theblixguy It seems like one way to approach it might be to extend a check for `isMetatypeBase` to also check for `functionRefKind != FunctionRefKind::Unapplied`

xedin commented:
Sounds good, thank you!

Commit:
[cs] use functionRefKind as well

theblixguy commented:
@xedin So I took a look at the test results today. I'm gonna focus on fixing the crashes first. So, there are some tests which are causing a crash because it's hitting the `unreachable` in `diagnoseUnviableLookupResults`.

1. `type member on instance is diagnosed via fixes`.

An example:

```swift
func enumMetatypeMember(_ opt: Int?) {
  opt.none // expected-error{{enum case 'none' cannot be used as an instance member}}
}
```

`AllowTypeOrInstanceMemberFailure::diagnoseAsError()` is never invoked in that example. In `performMemberLookup`, it gets added as `MemberLookupResult::UR_TypeMemberOnInstance`.

This is a bit strange because the `baseTy` isn't a `MetatypeType`. It's:

> (optional_type
>> (struct_type decl=Swift.(file).Int))

and the decl is:

> (enum_element_decl "none" interface type='<Wrapped> (Optional<Wrapped>.Type) -> Optional<Wrapped>' access=public)

The check `if (decl->isInstanceMember()) { ... }` isn't executed here.

2. `instance member on type is diagnosed via fixes`

An example (from `test/TypeCoercion/overload_member.swift`):

```
var clams : X

struct WeirdIvarLookupBehavior { 
  var clams : Y

  func f() {
    var _ : Y = clams
  }

  static func static_f() {
    // FIXME: These diagnostics still suck.
    var a : X = clams // expected-error{{member 'clams' cannot be used on type 'WeirdIvarLookupBehavior'}}
    var b : Y = clams // expected-error{{member 'clams' cannot be used on type 'WeirdIvarLookupBehavior'}}
  }
}
```

Same issue as (1): `AllowTypeOrInstanceMemberFailure::diagnoseAsError()` is never invoked in this example.

xedin commented:
@theblixguy Regarding the first one with optional - that's because of the `if (baseObjTy->getOptionalObjectType())` block - I'd suggest moving new code to the top to for it to properly diagnose things, it seems like it doesn't really need a second `performMemberLookup` anyway, and can diagnose using original `result`. I'd also suggest you wrap this into a static method something like `attemptToFixMissingOrExtraneousMetatype` which returns `true` if it fixed a problem and solve otherwise.

xedin commented:
I think problem with the last example is related to the fact that base is implicit. I see that without your changes it does add result to `UnviableCandidates` list.

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -3432,8 +3433,10 @@ performMemberLookup(ConstraintKind constraintKind, DeclName memberName,
+      if (((isa<FuncDecl>(decl) && !hasInstanceMethods) ||
+           (!isa<FuncDecl>(decl) && !hasInstanceMembers)) ||
+          (isMetatypeBase && !(functionRefKind == FunctionRefKind::Unapplied ||
+                               functionRefKind == FunctionRefKind::Compound))) {
```
How about we replace this whole `if` statement with this:

```c++
      // If this is a method and either base is not allowed to have any,
      // or it's not going to be used as a value e.g. `let _ = X.foo`, report.
      if (isa<FuncDecl>(decl)) {
        auto isUnappliedOrCompoundRef = [](FunctionRefKind FRK) {
          return FRK == FunctionRefKind::Unapplied ||
                 FRK == FunctionRefKind::Compound;
        };

        if (!hasInstanceMethods ||
            (isMetatypeBase && !isUnappliedOrCompoundRef())) {
          result.addUnviable(candidate,
                             MemberLookupResult::UR_InstanceMemberOnType);
          return;
        }
      // If this is a property and base type is not allowed to have any, report.
      } else if (!hasInstanceMembers) {
        result.addUnviable(candidate,
                           MemberLookupResult::UR_InstanceMemberOnType);
        return;
      }
```

theblixguy replied:
Let me try!

xedin replied:
Can you please push your latest code so I can pull and test it a bit tonight?

theblixguy replied:
Done. What I find strange is that it it's getting added as `UR_InstanceMemberOnType` but `AllowTypeOrInstanceMemberFailure::diagnoseAsError()` isn't getting called. Very interesting!

xedin replied:
One more problem is that new logic in`performMemberLookup` is not going to work, because Swift support partial application, maybe counter-intuitive but this is supposed to work:

```swift
struct S {
  func foo() {}
}

var s = S()
S.foo(s) // S.foo returns (S) -> () -> () which then gets partially applied on instance of S via `s`
```

xedin replied:
@theblixguy Here are some changes to your PR https://gist.github.com/xedin/b004412cdbbaabb278fe3b46a154c727

I need to think about the best way to diagnose these partial application problems:

```swift
class A {}

enum B {
  static func foo() {
    bar(A())
  }

  func bar(_: A) {}
}
```

Currently our constraint system is setup poorly for this kinds of things...

xedin replied:
@theblixguy I think the only way forward we have is it diagnose these partial application errors via fixes before we can merge this (it's good anyway because at least one related to mutable base is diagnosed by MiscDiagnostics at the moment), I'll look into doing that in a bit.

theblixguy replied:
@xedin Does that mean we need another `FailureDiagnostic` subclass, etc?

theblixguy replied:
@xedin Oh, perhaps it can be done inside ```if (attemptFixes && kind >= ConstraintKind::Conversion) { ... }```, but the meta type is already stripped off. Or maybe in ```case ConstraintKind::Equal: { ... }``` we can check if `type2` is a `MetatypeType` and we can apply a fix & return `getTypeMatchSuccess()`? If I do that (w/o the fix) then the code throws `cannot invoke 'g' with an argument list of type '(Bar)'`, but I don't think such a simple check is correct.

theblixguy replied:
Ah never mind, just saw your PR!

theblixguy replied:
@xedin Okay 😄

theblixguy replied:
@xedin now that the partial applications PR has been merged, should I resume working on this? or are there any other upcoming changes that you're working on that I need to wait for?

theblixguy replied:
@xedin Oh! Are you working on another PR to diagnose those situations? or should that be handled in this PR?

theblixguy replied:
@xedin Oh, okay! I have synced the PR with master, so we can run a test now!

theblixguy commented:
Thanks @xedin, moving the code to the top fixes the first problem - I'll extract the code into a separate static function. 

Regarding the second, is this due to the `(isMetatypeBase && !(functionRefKind == FunctionRefKind::Unapplied || functionRefKind == FunctionRefKind::Compound))` check? Not sure how I can check if the base is implicit or not as I don't have access to the base expression inside `performMemberLookup()`. Should I use `getContextualTypeNode()` or extract it from candidate

xedin commented:
@theblixguy I replied inline, I think it's because it was adding properties to unviable list because of new check.

Commit:
[cs] adds new if-check and moves unviable check to the top

Commit:
[cs] diagnose fixes in diagnoseUnviableLookupResults()

Commit:
[cs] remove check for metatype base, etc

xedin committed:
[CSFix] Introduce non-fatal "warning" fixes

The intention here is to be able to detect warnings earlier and
move some of the logic from CSApply and MiscDiagnostics to solver.

Warning fixes still lead to solution being applied to AST.

Commit:
Merge branch 'master' into fix/SR-9415

Commit:
Merge branch 'master' into fix/SR-9415

xedin commented:
@swift-ci please smoke test

Commit:
[cs] adds missing braces

theblixguy commented:
@xedin can you run the smoke test again? seems like the merge deleted the braces at the end of the some classes, causing a build failure 🤦‍♂️ should be fine now!

xedin commented:
@swift-ci please smoke test

xedin commented:
No problem :)

theblixguy commented:
@xedin I have updated the existing diagnostics - I think some of them are wrong, but could you take a look at the latest commit 52f367b?

Also, there's two crashes in `diagnoseUnviableLookupResults`, I'll look into them now.

xedin commented on test/Constraints/diagnostics.swift:
```diff
@@ -1031,7 +1031,6 @@ takesClosure() { ambiguousCall() } // expected-error {{ambiguous use of 'ambiguo
 class SR_4692_a {
   private static func foo(x: Int, y: Bool) {
     self.bar(x: x)
-    // expected-error@-1 {{instance member 'bar' cannot be used on type 'SR_4692_a'}}
```
What did be lose errors in this and `SR_4692_b` cases?

theblixguy replied:
I am not sure why we lost the diagnostic here, I'll look into this. I am a bit surprised that we've regressed further here as nothing's changed in my code since last time. Perhaps it's due to the recent changes to diagnostics code.

xedin reviewed (changes requested):
Looks like the most of the regressed diagnostics are related to `init` calls which should be easy to fix by porting them from CSDiag, other big problem is that we have lost fix-its in some cases which should be restored before we can merge this.

theblixguy commented:
@xedin I have moved the init diagnostic over and that fixes the regressed diagnostics. However, I am unsure about the lost fixits - it seems like we're not offering a fix-it if the lookup was [qualified](https://github.com/apple/swift/pull/21830/files#diff-34de70db160d5d09726050ab292a15acR1649):

```c++
// No fix-it if the lookup was qualified
if (expr && !expr->isImplicit()) {
    return true;
}
```

theblixguy commented:
Oh never mind, it seems like we should be using `getAnchor` instead of `getParentExpr` there. Then the fix-it shows up. @xedin

Commit:
[csdiagnostics] restore fix-its (hopefully) and update tests

Commit:
[csdiagnostics] update some more tests and fix fix-it loc

Commit:
[test] update some more tests

Commit:
[csdiagnostics] fix compiler crasher

Commit:
[csdiagnostics] fix another crash

theblixguy commented:
@xedin I have fixed all the crashes. At the moment, there's only 3 tests to fix.

1. `Constraints/diagnostics.swift`: No diagnostics for `SR_4692_a` & `SR_4692_b`.
2. `expr/postfix/dot/init_ref_delegation.swift`: Missing fixits
3. `NameBinding/name_lookup.swift`: Missing diagnostics for some cases. I am unsure about this one - it seems like `getResolvedOverload(getLocator())->Choice` doesn't return a choice with a `Decl` in some cases (for example, for function `staticTestSelf1` in `name_lookup.swift`). Do you know what's going on here?

xedin commented:
@theblixguy Great! Regarding #3 I'm not sure what exactly is going on there, I suspect that locator associated with the fix might not be complete in some cases, the easiest way to figure that out is to dump constraint system state and locator and see what are the locators associated with overload choices.

theblixguy commented:
Thanks @xedin, I'll give it a try now. I suspect it's due to subscripts - the crash log says the compiler was type checking a subscript (`super[0] = 42.0` for example). If I remove that test, then there's no more crashes.

xedin commented:
Then yes, most likely it's just that locator has to be adjusted to find that member properly.

theblixguy commented:
@xedin Hmm, so if I do `getLocator()->dump(&cs)`, then I get `locator@0x7f8b29aa1970 [Subscript@/Users/suyashsrijan/Desktop/test.swift:74:9 -> subscript member]`. So it seems like the locator is pointing to the subscript member.

xedin commented:
It's more important what overloads are associated with.

theblixguy commented:
@xedin Oh, dumping the CS gives:

> Score: 0 0 0 0 0 0 0 0 0 1 0
Type Variables:
  $T0 [lvalue allowed] as ThisDerived1.Type @ locator@0x7fb9e52d5400 [DeclRef@/Users/suyashsrijan/Desktop/test.swift:74:5]
  $T1 literal=3 bindings={(subtypes of) (default from ExpressibleByIntegerLiteral) Int} @ locator@0x7fb9e52d5480 [IntegerLiteral@/Users/suyashsrijan/Desktop/test.swift:74:10]
  $T2 [lvalue allowed] subtype_of_existential bindings={} @ locator@0x7fb9e52d5590 [Subscript@/Users/suyashsrijan/Desktop/test.swift:74:9 -> function result]
  $T3 as ($T4) -> $T5 @ locator@0x7fb9e52d5590 [Subscript@/Users/suyashsrijan/Desktop/test.swift:74:9 -> function result]
  $T4 subtype_of_existential bindings={} @ locator@0x7fb9e52d5600 [Subscript@/Users/suyashsrijan/Desktop/test.swift:74:9 -> subscript member -> function argument]
  $T5 [lvalue allowed] subtype_of_existential involves_type_vars bindings={} @ locator@0x7fb9e52d5600 [Subscript@/Users/suyashsrijan/Desktop/test.swift:74:9 -> subscript member -> function argument]
  $T6 subtype_of_existential involves_type_vars bindings={} @ locator@0x7fb9e52d5600 [Subscript@/Users/suyashsrijan/Desktop/test.swift:74:9 -> subscript member -> function argument]

> Active Constraints:

> Inactive Constraints:
  $T1 literal conforms to ExpressibleByIntegerLiteral [[locator@0x7fb9e52d5480 [IntegerLiteral@/Users/suyashsrijan/Desktop/test.swift:74:10]]];
  $T4 key path projecting ThisDerived1.Type -> $T5 [[locator@0x7fb9e52d5570 [Subscript@/Users/suyashsrijan/Desktop/test.swift:74:9 -> subscript member]]];
  $T5 equal $T6 [[locator@0x7fb9e52d5570 [Subscript@/Users/suyashsrijan/Desktop/test.swift:74:9 -> subscript member]]];
Resolved overloads:
  selected overload set choice key path application root ThisDerived1.Type
  selected overload set choice self: $T0 == ThisDerived1.Type

Seems like there's no decl because the type is a meta type?

theblixguy commented:
Seems like that fixed it - just had to look through the `MetatypeType` to get the underlying `ValueDecl`

Commit:
[csdiagnostics] look through metatype

theblixguy commented:
@xedin The reason why some of those fix-its weren't being generated in ` init_ref_delegation.swift` is because of this:

```
// No fix-it if the lookup was qualified
if (getAnchor() && !getAnchor()->isImplicit()) {
  return true;
}
```

If I get rid of this, then the fix-its are generated in _some_ cases, but not the others. This is because those diagnostics aren't emitted here (eg: `cannot_call_with_params` is emitted in `visitApplyExpr`). Not sure how this code broke those diagnostics, but I believe the solution would be to duplicate the diagnostic in `AllowTypeOrInstanceMemberFailure::diagnoseAsError()` as well?

And the missing diagnostics in some cases in `NameBinding/name_lookup.swift` isn't being generated because it's not getting detected as `UR_InstanceMemberOnType` or `UR_TypeMemberOnInstance`.

Commit:
[csdiagnostics] [test] update tests

xedin commented:
@theblixguy FYI, I think the reason why what you describe might happen are expression based diagnostics in CSDiag since while trying to re-typecheck sub-expressions are getting "fixed".

Also don't worry about `test/NameBinding/name_lookup.swift` I will get that fixed, it's related to partial application feature. Fixed that requires changing overload representation a bit to allow fixes to be attached early on.

xedin commented on test/expr/postfix/call/construction.swift:
```diff
@@ -113,7 +113,7 @@ protocol P2 {
 func constructExistentialValue(_ pm: P.Type) {
   _ = pm.init()
-  _ = P() // expected-error{{protocol type 'P' cannot be instantiated}}
+  _ = P() // expected-error{{instance member 'init' cannot be used on type 'P'}}
```
Old diagnostic was better here, see my comment in `test/Constraints/protocols.swift`

xedin commented on test/type/subclass_composition.swift:
```diff
@@ -510,8 +510,8 @@ func staticMembers(
 
-  _ = m2.instanceProtocolMember // expected-error {{instance member 'instanceProtocolMember' cannot be used on type 'ClassWithStaticMember & ProtocolWithStaticMember'}}
-  _ = m2.instanceClassMember // expected-error {{instance member 'instanceClassMember' cannot be used on type 'ClassWithStaticMember & ProtocolWithStaticMember'}}
+  _ = m2.instanceProtocolMember // expected-error {{static member 'instanceProtocolMember' cannot be used on instance of type '(ClassWithStaticMember & ProtocolWithStaticMember).Type'}}
```
Hm, this is incorrect, `instanceProtocolMember` is clearly not static and m{1, 2} are not an instance types...

xedin commented on test/expr/postfix/dot/init_ref_delegation.swift:
```diff
@@ -270,10 +270,10 @@ func foo<T: C>(_ x: T, y: T.Type) where T: P {
-  var ci2 = x.init(x: 0) // expected-error{{'init' is a member of the type; use 'type(of: ...)' to initialize a new object of the same dynamic type}} {{15-15=type(of: }} {{19-19=)}} 
-  var ci3 = x.init() // expected-error{{'init' is a member of the type; use 'type(of: ...)' to initialize a new object of the same dynamic type}} {{15-15=type(of: }} {{19-19=)}} 
-  var ci4 = x.init(proto: "") // expected-error{{'init' is a member of the type; use 'type(of: ...)' to initialize a new object of the same dynamic type}} {{15-15=type(of: }} {{19-19=)}} 
+  var ci1 = x.init(required: 0) // expected-error{{cannot invoke 'T.init' with an argument list of type '(required: Int)'}} {{15-15=type(of: }} {{19-19=)}} 
```
I think previous diagnostics in this block were better, since the actual problem is not related to arguments but to the base of the member reference...

theblixguy replied:
It doesn't get flagged as `UR_InstanceMemberOnType` or `UR_TypeMemberOnInstance`

xedin commented on test/Constraints/protocols.swift:
```diff
@@ -184,17 +184,17 @@ func staticExistential(_ p: P.Type, pp: P.Protocol) {
-  _ = P().bar(4) // expected-error{{protocol type 'P' cannot be instantiated}}
+  _ = P() // expected-error{{instance member 'init' cannot be used on type 'P'}}
+  _ = P().bar // expected-error{{instance member 'init' cannot be used on type 'P'}}
+  _ = P().bar(4) // expected-error{{instance member 'init' cannot be used on type 'P'}}
```
All these problems would benefit greatly from https://github.com/apple/swift/pull/22379. It would be great if we could expect the use of `AllowInvalidInitRef::onProtocolMetatype` introduced there to fix this case.

theblixguy replied:
How do we do that?

xedin replied:
I think it could be done in `simplifyMemberConstraint`. If this is a call on metatype and `baseTy->is<MetatypeType>() && baseTy->getMetatypeInstanceType()->isAnyExistentialType()` and the unviable declarations are `ConstructorDecl *`, it's equivalent to writing:

```swift
protocol P {}
_ = P() <- 
```

In that case you can record `AllowInvalidInitRef::onProtocolMetatype` fix and it should do the right thing.

theblixguy replied:
@xedin Thanks for your suggestion! `onProtocolMetatype` fixes this case 🎉 Can you take a look?

xedin commented on test/Constraints/protocols.swift:
```diff
@@ -210,8 +210,8 @@ func staticExistential(_ p: P.Type, pp: P.Protocol) {
-  _ = p.bar // expected-error{{instance member 'bar' cannot be used on type 'P'}}
-  _ = p.mut // expected-error{{instance member 'mut' cannot be used on type 'P'}}
+  _ = p.bar // expected-error{{static member 'bar' cannot be used on instance of type 'P.Type'}}
+  _ = p.mut // expected-error{{partial application of 'mutating' method is not allowed}} expected-error {{static member 'mut' cannot be used on instance of type 'P.Type'}}
```
`instance of` in the new error is incorrect, `P.Type` is a metatype...

xedin reviewed (changes requested):
I think it looks much better now! We just need a final push to fix the rest of the diagnostics...

Commit:
Merge branch 'master' into fix/SR-9415

Commit:
[csdiagnostics] [test] update test diagnostics

Commit:
[test] update diagnostics again

Commit:
[test] remove fixit which isnt emitted anymore

theblixguy commented:
@xedin I have fixed all crashes and incorrect diagnostics - could you take a look?

Commit:
[test] minor fixes

Commit:
[csdiagnostics] [test] add protocol fix

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -3762,21 +3762,42 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyMemberConstraint(
+        }
+
         auto *fix =
             AllowTypeOrInstanceMember::create(*this, baseTy, member, locator);
```
Can you try to simply form the overload set without producing this fix if base meets protocol metatype criteria, and choices are constructors (you can check the first one via `llvm::any_of` - if it's `ConstructorDecl` the rest are too). I think it should work just fine because there is `validateInitializerRef` now which is going to inroduce appropriate fixes.

theblixguy replied:
@xedin Done!

Commit:
[csdiagostics] simplify fix logic

theblixguy commented:
@xedin ah okay, all done now!

theblixguy commented:
Actually it leads to a crash in `CSApply`: `UNREACHABLE executed at /Users/suyashsrijan/Documents/swift-src/swift/lib/Sema/CSApply.cpp:6691` and causes a bunch of diagnostics to not be emitted.

theblixguy commented:
@xedin Seems like creating the `onProtocolMetatype` fix might be the only choice as perhaps the constraint system isn't set up in a way to deal with it automatically (as you mentioned `validateInitializerRef` would detect and apply the appropriate fixes, but it doesn't). I guess it's ok to create the fix for now! What do you think?

xedin commented:
@theblixguy I think it's worth to investigate why existing `AllowInvalidInitializerRef` fixes are not catching new cases, that's concerning to me.

theblixguy commented:
It might have something to do with the overload sets. So, for this example in `protocols.swift`:

```swift
func staticExistential(_ p: P.Type, pp: P.Protocol) {
  let _ = p()
  ...
}

I was able to make the diagnostic reappear by passing `baseTy` to `addOverloadSet()`. Of course we can't do this for all the cases, so maybe restrict it to when the base protocol metatype criteria is met.

theblixguy commented:
@xedin It might have something to do with the overload sets. So, for this example in `protocols.swift`:

```swift
func staticExistential(_ p: P.Type, pp: P.Protocol) {
  let _ = p()
  ...
}
```

I was able to make the diagnostic reappear by passing `baseTy` to `addOverloadSet()`. Of course we can't do this for all the cases, so maybe restrict it to when the base protocol metatype criteria is met.

xedin commented:
@theblixguy I think it has to be memberTy since that's what these choices are going to bind, the problem probably somewhere in `validateInitializerRef`, it looks like it might need a case for metatypes there since right now `onNonConstMetatype` handles everything besides.

theblixguy commented:
@xedin Oh, so if I add the case:

```c++
if (auto *MTT = baseType->getAs<MetatypeType>()) { ... }
else if (baseType->is<AnyMetatypeType>() && 
baseType->getMetatypeInstanceType()
        ->getWithoutParens()->isAnyExistentialType()) {
            (void)cs.recordFix(AllowInvalidInitRef::onProtocolMetatype(cs, baseType,
                           init, cs.isStaticallyDerivedMetatype(baseExpr), 
                           baseExpr->getSourceRange(), locator));
            return;
}
```
 then it works. But fails for this example:

```swift
protocol P {
  static func foo()
}

func staticExistential(_ p: P.Type, pp: P.Protocol) {
  _ = pp.foo
}
```

i.e. `validateInitializerRef` isn't called for this example and no diagnostic is emitted.

theblixguy commented:
Changing from && to || in ` if (!meetsProtocolBaseMetatypeCriteria || !requiresProtocolMetatypeFix) { ... }` fixes the issue though.

theblixguy commented:
@xedin It works now 🎉 Can you take a look?

Commit:
[cssimplify] adds onProtocolMetatype fix

xedin commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -1761,8 +1761,17 @@ static void validateInitializerRef(ConstraintSystem &cs, ConstructorDecl *init,
+               baseType->getMetatypeInstanceType()
+                   ->getWithoutParens()
+                   ->isAnyExistentialType()) {
+      (void)cs.recordFix(AllowInvalidInitRef::onProtocolMetatype(
```
You can integrate all of this into previous if statement. 

Also `P.Type` and `P.Protocol` types used as base for initializer should produce different errors - first should say that `.init` should be used explicitly like it does right now, the latter one should stay that protocol couldn't be instantiated.

theblixguy replied:
@xedin Done.

I am not sure about that, as for example:

```
let _ = p().bar(1)
```

in this case, the instance type is an existential type, so it doesn't meet the checks for `onNonConstMetatype` which is why this new diagnostic is emitted.

xedin commented on test/Constraints/protocols.swift:
```diff
@@ -178,19 +178,19 @@ func existential(_ p: P) {
-  let _ = p().bar(1) // expected-error{{initializing from a metatype value must reference 'init' explicitly}}
+  let _ = p() // expected-error{{value of type 'P.Type' is a protocol; it cannot be instantiated}}
+  let _ = p().bar // expected-error{{value of type 'P.Type' is a protocol; it cannot be instantiated}}
+  let _ = p().bar(1) // expected-error{{value of type 'P.Type' is a protocol; it cannot be instantiated}}
```
This diagnostic should stay the same, see my previous comment about difference between `P.Type` and `P.Protocol`.

xedin commented on test/Constraints/super_constructor.swift:
```diff
@@ -8,7 +8,7 @@ struct S {
 class D : B {
   func foo() {
-    super.init() // expected-error{{'super.init' cannot be called outside of an initializer}}
+    super.init() // expected-error{{member 'init' cannot be used on value of type 'B'}}
```
In this case new error is worse as well :(

Commit:
[cssimplify] simpify if check

Commit:
[cs] fix protocol diagnostics

theblixguy commented:
@xedin I have fixed the protocol diagnostics! Can you take a look?

Commit:
[cs] some renaming

xedin commented on test/Constraints/protocols.swift:
```diff
@@ -178,19 +178,19 @@ func existential(_ p: P) {
-  let _ = p() // expected-error{{value of type 'P.Type' is a protocol; it cannot be instantiated}}
-  let _ = p().bar // expected-error{{value of type 'P.Type' is a protocol; it cannot be instantiated}}
-  let _ = p().bar(1) // expected-error{{value of type 'P.Type' is a protocol; it cannot be instantiated}}
+  let _ = p() // expected-error{{initializing from a metatype value must reference 'init' explicitly}}
```
Great! Thank you! It would be great to extract this logic for `onNonConstMetatype` into separate lambda and try to simplify all of the conditions, that would de-dup a lot of code.

theblixguy replied:
@xedin Done! :)

xedin commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -1753,8 +1764,7 @@ static void validateInitializerRef(ConstraintSystem &cs, ConstructorDecl *init,
       if (!cs.isTypeReference(baseExpr) && instanceType->isExistentialType()) {
-        (void)cs.recordFix(AllowInvalidInitRef::onProtocolMetatype(
-            cs, baseType, init, true, baseExpr->getSourceRange(), locator));
+        recordOnProtocolMetatypeFix(/*isStaticallyDerived*/ true);
```
Sorry, not really what I meant, I was hopping that we could simplify the `if`s which leads to these fixes being produced...

theblixguy replied:
@xedin I have simplified the checks a more now, but I am not sure if it can be simplified any further.

xedin replied:
Ok, fair enough, that could be refactored later too! Let's get the rest of the regressed diagnostics under control (I think there is at least one which is related to super.init outside of initializer), and I think we should be ready to merge this PR.

theblixguy replied:
@xedin Okay! For that test, it doesn't get detected as `UR_InstanceMemberOnType` or `UR_TypeMemberOnInstance`. When I ran the tests, the unviable reasons were all different (most were type member on instance but others were label mismatch, etc).

xedin replied:
if the new logic doesn't get triggered why would that test case regress?

theblixguy replied:
@xedin Hmm, I think I am wrong. The fix does indeed get created, but the logic for emitting `could_not_use_value_member` is in `diagnoseUnviableLookupResults()` and not in the new logic 🤔 So it seems like the fix isn't getting diagnosed!

theblixguy replied:
Returning `SolutionKind::Error` in `simplifyMemberConstraint` as originally done if `!result.UnviableCandidates.empty()` works around the problem, but it's obviously not the correct thing to do as our fixes won't get created otherwise.

xedin replied:
I think it would make sense to either add new fix+diagnostic for this situation or incorporate the logic from CSDiag into `AllowTypeOrInstanceMemberFailure`.

theblixguy replied:
@xedin Hmm okay, I'll take a look tomorrow! Something's strange is going on as the fix is indeed being created, but not being diagnosed (i.e. nothing is calling `diagnoseAsError()` on it!). I am not sure how we're ending up in `diagnoseUnviableLookupResults()` if we've already recorded a fix.

Commit:
[cs] simplify checks further

Commit:
Merge branch 'master' into fix/SR-9415

Commit:
[cs] fix some diagnostics

Commit:
[cs] fix more diagnostics

Commit:
[test] fix indentation

Commit:
[cs] fix more diagnostics

theblixguy commented:
@xedin I have fixed all the `super` related diagnostics. Can you take a look? I'd like to create a separate PR to address this via fixes (don't want to bloat this PR any further).

xedin commented on test/NameBinding/name_lookup.swift:
```diff
@@ -506,14 +506,14 @@ class rdar27013358G<T> {
 class r23904262 {
   let x = 1
-  let y = x // expected-error {{cannot use instance member 'x' within property initializer; property initializers run before 'self' is available}}
+  let y = x // expected-error {{instance member 'x' cannot be used on type 'r23904262'}}
```
Looks like these diagnostics still need some work...

theblixguy replied:
@xedin Fixed!

xedin commented on test/ClangImporter/objc_parse.swift:
```diff
@@ -68,7 +68,7 @@ func classMethods(_ b: B, other: NSObject) {
   i += B.classMethod(1, with: 2)
 
-  i += b.classMethod() // expected-error{{static member 'classMethod' cannot be used on instance of type 'B'}}
+  i += b.classMethod() // expected-error{{member 'classMethod' cannot be used on value of type 'B'}}
```
Can you figure out why `classMethod` no longer diagnosed as `static member` here?

theblixguy replied:
@xedin Fixed!

xedin reviewed (commented):
Couple more diagnostics to figure out and we are good to do, thank you!

Commit:
[csdiagnostics] use rawAnchor() instead

theblixguy commented:
@xedin Fixed the remaining diagnostics, can you take a look?

xedin reviewed (approved):
LGTM!

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@theblixguy Should probably change the description to mention what this PR really does, because it doesn't fix [SR-9415](https://bugs.swift.org/browse/SR-9415) unfortunately.

theblixguy commented:
@xedin Done. Seems like that classMethod test case in `objc_parse.swift` is still failing, however I can't reproduce it locally with a simpler example:

```swift
class A {
  static func classMethod() -> Int { return 0 }
  class func anotherClassMethod() -> Int { return 1 }
}

func foo(_ a: A) {
  var i = A.classMethod()
  i += a.classMethod() // OK: static member 'classMethod' cannot be used on instance of type 'A'}}
  i += a.anotherClassMethod() // OK: static member 'anotherClassMethod' cannot be used on instance of type 'A'
}
```

I can't find the definition of `B` in the test method, seems like it's being imported from somewhere.

xedin commented:
@theblixguy That's a ClangImporter test, so `B` is most likely imported from Objective-C and located in one of the files in `test/ClangImporter/Inputs/custom-modules/`

theblixguy commented:
@xedin Oh, so its defined as

```objc 
+ (int)classMethod;
```

which is the equivalent of

```swift
class func classMethod() -> Int {}
```

However, I don't understand how the test case is producing a slightly different diagnostic where as my minimised example doesn't. Is it worth investigating or should I update the diagnostic? I mean, this works:

```swift
class Foo {
  class func bar() -> String {}
}

func doSomething(with f: Foo) {
  let _ = f.bar() // Ok: static member 'bar' cannot be used on instance of type 'Foo'
}
```

xedin commented:
I think it's worth investigating.

theblixguy commented:
@xedin Ok, so the reason is that the fix isn't being diagnosed - it does get created, but again, nothing is calling `diagnoseAsError()` on it. This is really strange as the recorded fix isn't being applied so I think something unrelated to this PR isn't working correctly. Any ideas? 

Some observations: I reduced the test file down to that one failing test case. I can see that the part after `if (shouldAttemptFixes()) { ... }` in `simplifyMemberConstraint()` is called 3 times and the fix is recorded once. Then, we end up in `diagnoseUnviableLookupResults()` where we end up emitting a `could_not_use_type_member` diagnostic instead.

theblixguy commented:
Interesting stack trace:

```
7  swiftc                   0x000000010864c24b (anonymous namespace)::FailureDiagnosis::diagnoseUnviableLookupResults(swift::constraints::MemberLookupResult&, swift::Expr*, swift::Type, swift::Expr*, swift::DeclName, swift::DeclNameLoc, swift::SourceLoc) + 1931
8  swiftc                   0x000000010865241e (anonymous namespace)::FailureDiagnosis::visitApplyExpr(swift::ApplyExpr*) + 12654
9  swiftc                   0x0000000108637316 swift::constraints::ConstraintSystem::diagnoseFailureForExpr(swift::Expr*) + 86
10 swiftc                   0x00000001086e0f87 swift::constraints::ConstraintSystem::salvage(llvm::SmallVectorImpl<swift::constraints::Solution>&, swift::Expr*) + 1319
11 swiftc                   0x0000000108692678 swift::constraints::ConstraintSystem::solve(swift::Expr*&, swift::Type, swift::ExprTypeCheckListener*, llvm::SmallVectorImpl<swift::constraints::Solution>&, swift::FreeTypeVariableBinding) + 72
12 swiftc                   0x000000010874945c swift::TypeChecker::typeCheckExpressionImpl(swift::Expr*&, swift::DeclContext*, swift::TypeLoc, swift::ContextualTypePurpose, swift::OptionSet<swift::TypeCheckExprFlags, unsigned int>, swift::ExprTypeCheckListener&, swift::constraints::ConstraintSystem*) + 908
13 swiftc                   0x00000001087490bf swift::TypeChecker::typeCheckExpression(swift::Expr*&, swift::DeclContext*, swift::TypeLoc, swift::ContextualTypePurpose, swift::OptionSet<swift::TypeCheckExprFlags, unsigned int>, swift::ExprTypeCheckListener*, swift::constraints::ConstraintSystem*) + 63
14 swiftc                   0x000000010863f3c2 (anonymous namespace)::FailureDiagnosis::typeCheckChildIndependently(swift::Expr*, swift::Type, swift::ContextualTypePurpose, swift::OptionSet<TCCFlags, unsigned int>, swift::ExprTypeCheckListener*, bool) + 1138
15 swiftc                   0x0000000108645e47 (anonymous namespace)::FailureDiagnosis::typeCheckArgumentChildIndependently(swift::Expr*, swift::Type, swift::CalleeCandidateInfo const&, swift::OptionSet<TCCFlags, unsigned int>) + 4487
16 swiftc                   0x0000000108651303 (anonymous namespace)::FailureDiagnosis::visitApplyExpr(swift::ApplyExpr*) + 8275
17 swiftc                   0x0000000108637316 swift::constraints::ConstraintSystem::diagnoseFailureForExpr(swift::Expr*) + 86
18 swiftc                   0x00000001086e0f87 swift::constraints::ConstraintSystem::salvage(llvm::SmallVectorImpl<swift::constraints::Solution>&, swift::Expr*) + 1319
19 swiftc                   0x0000000108692678 swift::constraints::ConstraintSystem::solve(swift::Expr*&, swift::Type, swift::ExprTypeCheckListener*, llvm::SmallVectorImpl<swift::constraints::Solution>&, swift::FreeTypeVariableBinding) + 72
20 swiftc                   0x000000010874945c swift::TypeChecker::typeCheckExpressionImpl(swift::Expr*&, swift::DeclContext*, swift::TypeLoc, swift::ContextualTypePurpose, swift::OptionSet<swift::TypeCheckExprFlags, unsigned int>, swift::ExprTypeCheckListener&, swift::constraints::ConstraintSystem*) + 908
21 swiftc                   0x00000001087490bf swift::TypeChecker::typeCheckExpression(swift::Expr*&, swift::DeclContext*, swift::TypeLoc, swift::ContextualTypePurpose, swift::OptionSet<swift::TypeCheckExprFlags, unsigned int>, swift::ExprTypeCheckListener*, swift::constraints::ConstraintSystem*) + 63
22 swiftc                   0x00000001087de93b swift::ASTVisitor<(anonymous namespace)::StmtChecker, void, swift::Stmt*, void, void, void, void>::visit(swift::Stmt*) + 587
23 swiftc                   0x00000001087ddb08 bool (anonymous namespace)::StmtChecker::typeCheckStmt<swift::BraceStmt>(swift::BraceStmt*&) + 136
24 swiftc                   0x00000001087dc678 swift::TypeChecker::typeCheckFunctionBodyUntil(swift::FuncDecl*, swift::SourceLoc) + 264
25 swiftc                   0x00000001087dd1cf swift::TypeChecker::typeCheckAbstractFunctionBody(swift::AbstractFunctionDecl*) + 271
26 swiftc                   0x00000001087fe78e typeCheckFunctionsAndExternalDecls(swift::SourceFile&, swift::TypeChecker&) + 286
27 swiftc                   0x00000001087ff3e9 swift::performTypeChecking(swift::SourceFile&, swift::TopLevelContext&, swift::OptionSet<swift::TypeCheckingFlags, unsigned int>, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int) + 1065
28 swiftc                   0x0000000107f9c456 swift::CompilerInstance::parseAndTypeCheckMainFileUpTo(swift::SourceFile::ASTStage_t, swift::PersistentParserState&, swift::DelayedParsingCallbacks*, swift::OptionSet<swift::TypeCheckingFlags, unsigned int>) + 502
29 swiftc                   0x0000000107f9ae33 swift::CompilerInstance::parseAndCheckTypesUpTo(swift::CompilerInstance::ImplicitImports const&, swift::SourceFile::ASTStage_t) + 339
30 swiftc                   0x0000000107f9a517 swift::CompilerInstance::performSemaUpTo(swift::SourceFile::ASTStage_t) + 615
31 swiftc                   0x0000000107cf068d performCompile(swift::CompilerInstance&, swift::CompilerInvocation&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) + 1613
32 swiftc                   0x0000000107ceefe6 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 2982
33 swiftc                   0x0000000107c96032 main + 690
34 libdyld.dylib            0x00007fff78e6bed9 start + 1
```

So it seems like we're diagnosing visiting `visitApplyExpr()` twice and trying to solve and salvage and then end up in `diagnoseUnviableLookupResults()`

xedin commented:
@theblixguy Are you sure that `diagnoseAsError` is not getting called or is it a case where that method returns `false` which then triggers old expression based diagnostics to kick-in?

theblixguy commented:
@xedin I put a `llvm_unreachable()` inside `diagnoseAsError()` but that was never triggered! It's only happening for this specific case, I can't reproduce it with any other test case.

xedin commented:
I think it would be fine if you fix it in `diagnoseUnviableLookupResults` to diagnose properly and fix in the follow-up PR.

theblixguy commented:
@xedin I tried creating a `AllowTypeOrInstanceMemberFailure` fix before we emit the `could_not_use_type_member` so I can diagnose this test case properly. It works however it leads to incorrect diagnostics in other situations and sometimes a crash possibly due to wrong overloads. For this specific test case, the unviable reasons aren't all the same.

I dumped the CS before emitting the diagnostic:

```
Score: 0 0 0 0 0 0 0 0 0 0 0
Type Variables:
  $T0 [lvalue allowed] as B @ locator@0x7f869a3b2618 [DeclRef@/Users/suyashsrijan/Documents/swift-src/swift/test/ClangImporter/objc_parse.swift:71:8]
  $T1 [lvalue allowed] as () -> $T2 @ locator@0x7f869a3b2698 [UnresolvedDot@/Users/suyashsrijan/Documents/swift-src/swift/test/ClangImporter/objc_parse.swift:71:10 -> member]
  $T2 #defaultable_bindings=1 bindings={Any} @ locator@0x7f869a3b2740 [Call@/Users/suyashsrijan/Documents/swift-src/swift/test/ClangImporter/objc_parse.swift:71:10 -> function result]

Active Constraints:

Inactive Constraints:
  $T2 can default to Any [[locator@0x7f869a3b2698 [UnresolvedDot@/Users/suyashsrijan/Documents/swift-src/swift/test/ClangImporter/objc_parse.swift:71:10 -> member]]];
Resolved overloads:
  selected overload set choice b: $T0 == B


Failed constraint:
  B[.classMethod: value] == $T1 [[locator@0x7f869a3b2698 [UnresolvedDot@/Users/suyashsrijan/Documents/swift-src/swift/test/ClangImporter/objc_parse.swift:71:10 -> member]]];
```

I can either try creating the diagnostic directly (hopefully it won't trigger in other scenarios) or I can update the test file for now and update it back in a follow-up PR.

xedin commented:
Based on the state of constraint system, it doesn't look like `AllowTypeOrInstanceMemberFailure` fix is not created for that situation since it just records failed constraint?...

theblixguy commented:
@xedin It is indeed created and recorded, just not diagnosed (unsure why). It only happens for this specific case - I tried other variants but can't reproduce. The unviable reasons are:

```
UR_LabelMismatch
UR_LabelMismatch
UR_TypeMemberOnInstance
UR_LabelMismatch
UR_TypeMemberOnInstance
UR_TypeMemberOnInstance
UR_TypeMemberOnInstance
UR_TypeMemberOnInstance
```

Also, the `member` we have isn't static, so I can't actually create a `could_not_use_type_member_on_instance` diagnostic for that case there.

xedin commented:
I guess the only question I have is - if the fix is emitted, why doesn't it get diagnosed, is there a bug elsewhere? I think this case shouldn't require anything special and should be handled by the failure you've added.

theblixguy commented:
@xedin Likely, although I am not sure if it's a bug in my PR or in general. I poked around a bit, it seems like we're ending up in `diagnoseMethodAttributeFailures()` which is calling `diagnoseUnviableLookupResults()`.

So our base type `(class_type decl=ObjectiveC.(file).B)` and base expression:

> (declref_expr type='<null>' decl=objc_parse.(file).classMethods(_:other:).b@/Users/suyashsrijan/Documents/swift-src/swift/test/ClangImporter/objc_parse.swift:66:21 function_ref=unapplied)

and the expression:

> (unresolved_dot_expr type='<null>' field 'classMethod' function_ref=double
  (declref_expr type='<null>' decl=objc_parse.(file).classMethods(_:other:).b@/Users/suyashsrijan/Documents/swift-src/swift/test/ClangImporter/objc_parse.swift:66:21 function_ref=unapplied))

xedin commented:
What happens in `simplifyMemberConstraint`? Does name lookup actually produce unviable results for `b.classMethod()` and what is the failure kind in there? 

I think that's more important to figure out why expression diagnostics got run in the first place...

theblixguy commented:
@xedin Yes. So the code in `if (!result.UnviableCandidates.empty()) { .. }` is called 3 times. We then do a check to ensure the unviable reasons for all the candidates are the same (all `UR_InstanceMemberOnType` or `UR_TypeMemberOnInstance`). We get a match 1 time, so we create and record a fix and return the overload set.

The unviable reasons for 3 lookups are:

1.
```
UR_LabelMismatch
UR_LabelMismatch
UR_TypeMemberOnInstance
UR_LabelMismatch
UR_TypeMemberOnInstance
UR_TypeMemberOnInstance
UR_TypeMemberOnInstance
UR_TypeMemberOnInstance
```

2.
```
UR_LabelMismatch
UR_LabelMismatch
UR_TypeMemberOnInstance
UR_LabelMismatch
UR_TypeMemberOnInstance
UR_TypeMemberOnInstance
UR_TypeMemberOnInstance
UR_TypeMemberOnInstance
```

3. (this matches our criteria so we create a fix)
```
UR_TypeMemberOnInstance
UR_TypeMemberOnInstance
UR_TypeMemberOnInstance
UR_TypeMemberOnInstance
```

xedin commented:
Ok, I think I understand what is going on here now, thank you! Maybe the best solution would be to filter out anything which isn't `UR_TypeMemberOnInstance` or `UR_InstanceMemberOnType` from choices before forming new overload set, what to give it a try?

So the logic is going to be not to check if there are any `UR_TypeMemberOnInstance` or `UR_InstanceMemberOnType` but instead just collect them, and if that list is not-empty create a fix and try the overloads.

Once this lands I also want to try and change the way we add overloads and make it less expensive, but that could be done separately.

Commit:
[cssimplify] filter unviable reasons

theblixguy commented:
@xedin That works! Can you run the CI?

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -3768,14 +3768,19 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyMemberConstraint(
+      SmallVector<std::pair<OverloadChoice, MemberLookupResult::UnviableReason>,
+                  4>
+          filteredURs;
+      llvm::copy_if(
```
`filteredURs` and `choices` could be a single collection there is no need to copy data back and forth.

theblixguy replied:
@xedin Done!

Commit:
[cssimplify] remove choices array

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

theblixguy commented:
@xedin Seems like this the filter approach isn't going to work as there's 7 test failures :(

xedin commented:
@theblixguy Looks like most of the diagnostics are improvements like:

`test/decl/protocol/protocol_with_superclass_where_clause.swift`
`test/decl/protocol/protocol_with_superclass.swift`
and 
`test/Generics/function_defs.swift`

The rest are related to `init` use, used to be diagnosed by expression diagnostics before and now are diagnosed via new fixes, so diagnostic has to be updated to support that e.g. https://github.com/apple/swift/pull/21830/files#diff-5da71108df45946c4eba89a9cc3a323bR865
doesn't belong in CSDiag.cpp anymore.

Commit:
Merge branch 'master' into fix/SR-9415

Commit:
[test] update test diagnostics

theblixguy commented:
@xedin Oh! I have updated those diagnostic messages to the improved one. I'll fix the `init` ones now. The reason I put it there in `CSDiag.cpp` is because were would end up in `diagnoseUnviableLookupResults` for that test case as well (i.e. it had the same issue as the example we were discussing earlier). I guess the filtering approach may have fixed it, so I'll take a look now.

Commit:
Merge branch 'master' into fix/SR-9415

Commit:
[csdiag] remove init ref code

theblixguy commented:
@xedin I have removed the init ref code from CSDiag `diagnoseUnviableLookupResults`, however there's still one issue where the following code emits the wrong diagnostic:

```
class B {
  init() {}
  init(x: Int) {}
}

class D : B {
  override init() {
    super.init()
  }

  func super_calls() {
    // expected: 'super.init' cannot be called outside of an initializer
    // actual: missing argument label 'x:' in call
    super.init(0)
  }
}
```

So, dumping the CS makes it clear that the fix is created (actually twice):

```
Score: 1 0 0 0 0 0 0 0 0 0 0
Type Variables:
  $T0 [lvalue allowed] subtype_of_existential bindings={} @ locator@0x10a043618 [UnresolvedDot@/Users/suyashsrijan/Desktop/test.swift:14:11 -> member]

Active Constraints:

Inactive Constraints:

Fixes:
  [fix: allow access to instance member on type or a type member on instance] @ locator@0x10a043618 [UnresolvedDot@/Users/suyashsrijan/Desktop/test.swift:14:11 -> member]

-----

Score: 1 0 0 0 0 0 0 0 0 0 0
Type Variables:
  $T0 [lvalue allowed] subtype_of_existential bindings={} @ locator@0x10a04dc00 [UnresolvedDot@/Users/suyashsrijan/Desktop/test.swift:14:11 -> member]

Active Constraints:

Inactive Constraints:

Fixes:
  [fix: allow access to instance member on type or a type member on instance] @ locator@0x10a04dc00 [UnresolvedDot@/Users/suyashsrijan/Desktop/test.swift:14:11 -> member]
```

but again, it's not getting diagnosed (i.e. nothing is called diagnoseAsError() on the fix) and we end up calling `diagnoseParameterErrors()`.

xedin commented:
It looks like it might be producing ambiguous solution judging by two solutions you mentioned but it seemingly happening during expression diagnostics, because there is only one type variable which for regular call is too few. I'll checkout your code and try to run it to see what is going on.

xedin commented:
Ah I see, this label problem does not get diagnosed via fix, that's why it complains about it first, once that's fixed there is going to be `'super.init' cannot be called outside of an initializer` produced. 

Let me see if I can fix that to get label problem diagnosed via fix so we can get both diagnostics at the same time.

xedin commented:
@theblixguy I've opened https://github.com/apple/swift/pull/22797 to address this problem. Found one more simplication last moment, re-running tests locally to see if it works, and then I'm going to run tests on that PR.

theblixguy commented:
@xedin Great! Once your PR gets merged, I'll add the label diagnostic to that test case, run tests & and push my changes. Hopefully this will be the last one🤞

xedin commented:
Sounds like a plan! :)

Commit:
Merge branch 'master' into fix/SR-9415

Commit:
[test] add missing argument label diagnostic

theblixguy commented:
@xedin I have updated the test diagnostic - ran tests locally and everything passes 🎉

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

theblixguy commented:
All tests have passed @xedin 🎉 🎉 🎉

xedin commented:
Great! @theblixguy I'm going to squash everything together before merging if you don't mind? A lot of commit in the branch are merge commits and indentation changes.

theblixguy commented:
No problem @xedin!

xedin commented:
Thanks again for sticking with this @theblixguy!

theblixguy commented:
No problem, thanks a lot for your guidance and help as well @xedin :)
